File: Homework1/Codes/22202709_Ozongun_HalilArda.go
package main

import "fmt"
import "reflect"

func main() {

	fmt.Println("1: Legal types for subscripts: ")
	arr1 := [5]int{1,2,3,4,5}
	fmt.Println(arr1[3])
	var index int64 = 4
	fmt.Println(arr1[index])
	fmt.Println(arr1[int64(3)]) 
	//fmt.Println(arr1[3.3]) // Compile time Error: invalid array index 3.3 (type float64)
	fmt.Println(arr1[3.0])
	// fmt.Println(arr1["3"]) // Compile time Error: invalid array index "3" (type string)
	var index3 float64 = 3.0
	fmt.Println(arr1[int(index3)])
	// fmt.Println(arr1[index3]) // Compile time Error: (variable of type float64) must be integer



	fmt.Println("2: Range checking in Subscripting expressions: ")
	// fmt.Println(arr1[9]) // Compile time error for constant index out of range
	// fmt.Println(arr1[-2]) // Compile time error for negative index
	//var index2 = 9
	// fmt.Println(arr1[index2]) // Runtime error: index out of range [9] with length 5


	
	fmt.Println("3: Ragged multidimensional arrays: ")
	
	var multidimensionalArr = [3][2]int{{1,2}, {3,4}, {6,7}}
	//var multidimensionalArrOOB = [3][2]int{{1,2}, {3,4}, {6,7,8}} // Compile time error.
	fmt.Println(multidimensionalArr)

	fmt.Println(reflect.TypeOf(multidimensionalArr).Kind())
	fmt.Println(reflect.TypeOf(multidimensionalArr[0]).Kind())

	var raggedArr = [3][4]int{{1,2,3,4}, {5,6}, {7,8,9}} // Legal. but not ragged. It fills with 0
	fmt.Println(raggedArr)
	fmt.Println(reflect.TypeOf(raggedArr).Kind())
	fmt.Println(reflect.TypeOf(raggedArr[1]).Kind())

	var raggedSlice = [3][]int{{1,2}, {3,4,5}, {6,7,8,9}} // If size is not given, it is array of slices
	fmt.Println(raggedSlice)
	fmt.Println(reflect.TypeOf(raggedSlice).Kind())
	fmt.Println(reflect.TypeOf(raggedSlice[1]).Kind())


	fmt.Println("4: Initialization of arrays: ")
	
	var arr2 = [5]int{1,2,3,4,5}
	fmt.Println(reflect.TypeOf(arr2).Kind())

	var arr3 = [...]int{1,2,3,4,5}
	fmt.Println(reflect.TypeOf(arr3).Kind())

	var arr4 = [5]int{} // Initialization with only size

	fmt.Println(arr4)
	fmt.Println(reflect.TypeOf(arr4).Kind())

	var arr5 [5]int = [5]int{1,2,3,4,5}
	fmt.Println(reflect.TypeOf(arr5).Kind())
	



	fmt.Println("5: Array Slice: ")
	slice1 := arr1[1:3] // Slice of array
	slice2 := arr1[1:] // Slice of array with capacity
	fmt.Println(slice1)
	fmt.Println(reflect.TypeOf(slice1).Kind())
	fmt.Println(slice2)
	fmt.Println(reflect.TypeOf(slice2).Kind())
	// slice3 := arr1[1:3:1] //  Compile time error.



	
	fmt.Println("6: Allowed Operations: ")

	a := [3]int{1, 2, 3}
	b := [3]int{4, 5, 6}
	c := [3]int{1, 2, 3}
	
	fmt.Println("Equality: value-based")
	fmt.Println("a == b:", a == b) 
	fmt.Println("a == c:", a == c)
	
	/*d := [4]int{1, 2, 3, 4}
	fmt.Println("a == d:", a == d)*/ // Compile-time error: mismatched types
	
	fmt.Println("Adress equality using pointers")
	p1 := &a
	p2 := &b
	fmt.Println("p1 == p2:", p1 == p2)
	
}

File: Homework1/Codes/22202709_Ozongun_HalilArda.rs
use std::any::type_name;
use std::convert::TryInto;
use std::ptr::eq;


fn type_of<T>(_: &T) -> &'static str {
    type_name::<T>()
}

fn print_array(arr: &[i32]) {
    for i in 0..arr.len() {
        print!("{}  ", arr[i]);
    }
    println!(" ");
}

fn main(){
    println!("1: Legal types for subscripts:");
    let arr1 = [1, 2, 3, 4, 5];
    println!("{}", arr1[0]);
    // println!("{}", arr1["0"]); // Compile time error: error[E0277]: the type `[{integer}]` cannot be indexed by `&str`
    //println!("{}", arr1[0.0]); // Compile time error: error[E0277]: the type `[{integer}]` cannot be indexed by `float`
    // println!("{}", arr1[0.0f32]); // Compile time error: error[E0277]: the type `[{integer}]` cannot be indexed by `f32`
    let mut i = 0;
    println!("{}", arr1[i]);
    println!("type: of i: {}", type_of(&i));
    let i2: u32 = 0;
    // println!("{}", arr1[i2]); //Compile time errorL error[E0277]: the type `[{integer}]` cannot be indexed by `u32`




    println!("\n\n2: Range checking in Subscripting expressions: ");
    //println!("{}", arr1[9]);  // compile-time denial -> Don't compile.
    //println!("{}", arr1[-3]); // Compile time error. Negative integers cannot be used to index on a [{integer},5]
    i = 98;
    // println!("{}", arr1[i]); // Run-time error.




    println!("\n\n3: Ragged multidimensional arrays: ");
    let multidimensionalArr = [[1, 2, 3], [4, 5, 6]];
    println!("type: of multidimensionalArr: {}", type_of(&multidimensionalArr));

    //let raggedArr = [[1, 2, 3], [4, 5]]; // Compile time error.
    //println!("type: of raggedArr: {}", type_of(&raggedArr)); 

    // array of vectors
    let arr_of_vec = [vec![1, 2, 3], vec![4, 5, 6]];
    println!("type: of arr_of_vec: {}", type_of(&arr_of_vec));
    
    



    println!("\n\n4: Initialization of arrays: ");
    let mut arr2 = [5, 6, 7, 8, 9]; // Mutable array
    let arr3 = [5, 6, 7, 8, 9]; // Immutable array
    let arr4 = [6; 5];
    print_array(&arr4);
    let arr6: [i32; 5] = [45263, 45263, 45263, 45263, 45263];
    print_array(&arr6);
    let vec = vec![43, 45, 67, 89, 90]; // Vector
    let arr7: [i32; 5] = vec.try_into().unwrap();
    print_array(&arr7);
    



    println!("\n\n5: Array Slices: ");
    let slice = &arr1[1..4]; // 1 to 3

    println!("slice: {:?}", slice);
    println!("type of slice: {}", type_of(&slice));

    let slice2 = &arr1[1..];
    println!("slice2: {:?}", slice2);
    println!("type of slice2: {}", type_of(&slice2));
    // step slicing is not supported



    println!("\n\n6: Allowed Operations: ");

    let a = [1, 2, 3];
    let b = [4, 5, 6];
    let c = [1, 2, 3];
    println!("a: {:?}, b: {:?}, c: {:?}", a, b, c);

    println!("Compare contents:");
    println!("a == b: {}", a == b);
    println!("a == c: {}", a == c);
    println!("a != c: {}", a != c); 
    println!("a < b: {}", a < b); 

    println!("Compare references:");
    println!("eq(&a, &b): {}", eq(&a, &b));
    println!("eq(&a, &c): {}", eq(&a, &c));
    let d = &a;
    println!("eq(&a, d): {}", eq(&a, d));
       
}
    